In [1]:
# import required libraries
import pandas as pd
import numpy as np

In [2]:
# Loading the dataset
data = pd.read_csv("C:\\Users\\arcom\\Downloads\\kaggle download\\dirty_cafe_sales.csv")

We start by previewing and understanding the structure of the dataset

In [4]:
# Structure of the dataset
data.head(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2            2.0         4.0   
1    TXN_4977031      Cake        4            3.0        12.0   
2    TXN_4271903    Cookie        4            1.0       ERROR   
3    TXN_7034554     Salad        2            5.0        10.0   
4    TXN_3160411    Coffee        2            2.0         4.0   
5    TXN_2602893  Smoothie        5            4.0        20.0   
6    TXN_4433211   UNKNOWN        3            3.0         9.0   
7    TXN_6699534  Sandwich        4            4.0        16.0   
8    TXN_4717867       NaN        5            3.0        15.0   
9    TXN_2064365  Sandwich        5            4.0        20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

In [5]:
data.shape

(10000, 8)

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [12]:
# Checking null values in the dataset
data.isnull().sum()

Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

In [14]:
# Checking for duplicate values
data.duplicated().sum()

0

- The dataset contains **10,000 rows and 8 columns**.  
- Numerical columns like *Quantity*, *Price Per Unit*, and *Total Spent* are stored as **object types** instead of numeric.  
- Several columns have **missing values**, especially:
  - Payment Method → 2,579 missing  
  - Location → 3,265 missing
- There are no duplicates in the dataset.

**Next Steps:**  
Convert numeric columns and date time column to the correct data types and handle missing values before analysis.  

In [17]:
# Convert specific columns to numeric
data["Quantity"] = pd.to_numeric(data["Quantity"], errors="coerce")
data["Price Per Unit"] = pd.to_numeric(data["Price Per Unit"], errors="coerce")
data["Total Spent"] = pd.to_numeric(data["Total Spent"], errors="coerce")

In [19]:
# Convert specific columns to datetime
data["Transaction Date"] = pd.to_datetime(data["Transaction Date"], errors="coerce")

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9667 non-null   object        
 2   Quantity          9521 non-null   float64       
 3   Price Per Unit    9467 non-null   float64       
 4   Total Spent       9498 non-null   float64       
 5   Payment Method    7421 non-null   object        
 6   Location          6735 non-null   object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


In [23]:
data.head(10)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373    Coffee       2.0             2.0          4.0   
1    TXN_4977031      Cake       4.0             3.0         12.0   
2    TXN_4271903    Cookie       4.0             1.0          NaN   
3    TXN_7034554     Salad       2.0             5.0         10.0   
4    TXN_3160411    Coffee       2.0             2.0          4.0   
5    TXN_2602893  Smoothie       5.0             4.0         20.0   
6    TXN_4433211   UNKNOWN       3.0             3.0          9.0   
7    TXN_6699534  Sandwich       4.0             4.0         16.0   
8    TXN_4717867       NaN       5.0             3.0         15.0   
9    TXN_2064365  Sandwich       5.0             4.0         20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

- Converted **Quantity**, **Price Per Unit**, and **Total Spent** from `object` → numeric.
-  Converted **Transaction Date** from `object` → `datetime` for easier time-based analysis.
Now the dataset is in the correct format for calculations and trend analysis. This ensures accurate results when we perform aggregations.
  
**Next Step**  
Now that the data types are fixed, the next task is to address the missing values 

In [26]:
data['Item'].unique()

array(['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'UNKNOWN',
       'Sandwich', nan, 'ERROR', 'Juice', 'Tea'], dtype=object)

In [28]:
data[['Item','Price Per Unit']].value_counts()

Item      Price Per Unit
Juice     3.0               1110
Coffee    2.0               1108
Cake      3.0               1085
Sandwich  4.0               1082
Salad     5.0               1082
Smoothie  4.0               1036
Cookie    1.0               1026
Tea       1.5               1023
ERROR     3.0                 77
UNKNOWN   3.0                 77
          4.0                 70
ERROR     4.0                 61
UNKNOWN   2.0                 49
          1.0                 45
          5.0                 45
          1.5                 40
ERROR     5.0                 39
          1.5                 37
          1.0                 34
          2.0                 31
Name: count, dtype: int64

Interestingly, we also spotted unusual entries like **ERROR** and **UNKNOWN**, which seem to be data quality issues. These will need cleaning before we can trust the insights.  
We will replace these with `NaN` so they can be treated as missing data. This keeps the dataset clean and ensures our analysis reflects only real cafe sales.  

In [31]:
# replace error and unknown with nan
data.replace(['UNKNOWN','ERROR'],np.nan,inplace=True)

In [33]:
data['Price Per Unit'] = data['Price Per Unit'].fillna(data['Total Spent'] / data['Quantity'])
data['Quantity'] = data['Quantity'].fillna(data['Total Spent'] / data['Price Per Unit'])
data['Total Spent'] = data['Total Spent'].fillna(data['Price Per Unit'] * data['Quantity'])
                                                       

In [35]:
data.isnull().sum()

Transaction ID         0
Item                 969
Quantity              38
Price Per Unit        38
Total Spent           40
Payment Method      3178
Location            3961
Transaction Date     460
dtype: int64

Some transactions were missing prices, quantities, or totals. Since these three columns are related, we used simple math to fill the gaps:  
- If **Price Per Unit** was missing, we calculated it as `Total Spent ÷ Quantity`.  
- If **Quantity** was missing, we estimated it as `Total Spent ÷ Price Per Unit`.  
- If **Total Spent** was missing, we derived it as `Price Per Unit × Quantity`.
- We can see that most of the numeric gaps on the column Quantity, Item and Price Per Unit have been filled

This way, we ensured that all transactions had complete and consistent values without guessing randomly.  

In [38]:
df =data.groupby('Item')['Price Per Unit'].unique()
df

Item
Cake        [3.0, nan]
Coffee      [2.0, nan]
Cookie      [1.0, nan]
Juice       [3.0, nan]
Salad       [5.0, nan]
Sandwich    [4.0, nan]
Smoothie    [4.0, nan]
Tea         [1.5, nan]
Name: Price Per Unit, dtype: object

In [40]:
data[data['Item'].isnull()].head(10)

Transaction ID Item  Quantity  Price Per Unit  Total Spent  Payment Method  \
6     TXN_4433211  NaN       3.0             3.0          9.0             NaN   
8     TXN_4717867  NaN       5.0             3.0         15.0             NaN   
14    TXN_8915701  NaN       2.0             1.5          3.0             NaN   
30    TXN_1736287  NaN       5.0             2.0         10.0  Digital Wallet   
31    TXN_8927252  NaN       2.0             1.0          2.0     Credit Card   
33    TXN_7710508  NaN       5.0             1.0          5.0            Cash   
36    TXN_6855453  NaN       4.0             3.0         12.0             NaN   
52    TXN_8914892  NaN       5.0             5.0         25.0  Digital Wallet   
61    TXN_8051289  NaN       1.0             3.0          3.0             NaN   
63    TXN_9099694  NaN       3.0             5.0         15.0             NaN   

    Location Transaction Date  
6   Takeaway       2023-10-06  
8   Takeaway       2023-07-28  
14  In-store       2023-03-21  
30       NaN       2023-06-02  
31       NaN       2023-11-06  
33       NaN              NaT  
36  In-store       2023-07-17  
52       NaN       2023-03-15  
61  In-store       2023-10-09  
63  Takeaway       2023-11-18

In [42]:
# Creating a dictionary for Price Per Unit values to their corresponding Items
item_price = { 2.0 : 'Coffee',
               1.0 : 'Cookie',
               5.0 : 'Salad',
               1.5 : 'Tea'}
# Applying a function to fill missing Item values based on Price Per Unit
data['Item'] = data.apply(lambda row : item_price[row['Price Per Unit']]
              if (pd.isna(row['Item'])) and row['Price Per Unit'] in item_price
              else row['Item'], axis=1)

In [44]:
data[data['Price Per Unit'].isnull()].head(20)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
65      TXN_4987129  Sandwich       3.0             NaN          NaN   
629     TXN_9289174      Cake       NaN             NaN         12.0   
912     TXN_1575608  Sandwich       NaN             NaN         20.0   
1008    TXN_7225428       Tea       NaN             NaN          3.0   
1436    TXN_7590801       Tea       NaN             NaN          6.0   
1482    TXN_3593060  Smoothie       NaN             NaN         16.0   
1674    TXN_9367492       Tea       2.0             NaN          NaN   
1761    TXN_3611851       NaN       4.0             NaN          NaN   
2229    TXN_8498613  Sandwich       2.0             NaN          NaN   
2289    TXN_7524977       NaN       4.0             NaN          NaN   
2330    TXN_3849488     Salad       NaN             NaN          5.0   
2585    TXN_1259340       Tea       3.0             NaN          NaN   
3162    TXN_3577949      Cake       3.0             NaN          NaN   
3598    TXN_2857444  Smoothie       1.0             NaN          NaN   
3635    TXN_6177081    Cookie       NaN             NaN          1.0   
3673    TXN_6500126  Smoothie       2.0             NaN          NaN   
3779    TXN_7376255       NaN       NaN             NaN         25.0   
4021    TXN_6424202    Cookie       2.0             NaN          NaN   
4152    TXN_9646000       NaN       2.0             NaN          NaN   
5639    TXN_6206792       Tea       NaN             NaN          6.0   

      Payment Method  Location Transaction Date  
65               NaN  In-store       2023-10-20  
629   Digital Wallet  In-store       2023-12-30  
912              NaN  Takeaway       2023-01-05  
1008     Credit Card  Takeaway       2023-03-07  
1436            Cash  Takeaway              NaT  
1482            Cash       NaN       2023-03-05  
1674            Cash  In-store       2023-06-19  
1761     Credit Card       NaN       2023-02-09  
2229             NaN       NaN       2023-11-08  
2289             NaN       NaN       2023-12-09  
2330             NaN  In-store       2023-03-01  
2585  Digital Wallet       NaN       2023-02-24  
3162             NaN  Takeaway       2023-04-25  
3598            Cash  Takeaway       2023-05-10  
3635            Cash       NaN       2023-07-26  
3673             NaN       NaN              NaT  
3779             NaN  In-store       2023-05-27  
4021     Credit Card  In-store       2023-11-20  
4152             NaN  In-store       2023-12-14  
5639     Credit Card       NaN       2023-10-13

Since the prices for each product are consistent, the missing values can be directly filled with the known price per item.  
This ensures the dataset remains accurate, complete, and reliable for further analysis.

In [47]:
# Creating the dictionary 
price_to_item = {'Cake': 3.0 , 'Coffee': 2.0 ,
                 'Cookie': 1.0,'Juice': 3.0,
                 'Salad': 5.0,'Sandwich': 4.0,
                 'Smoothie': 4.0,'Tea': 1.5}
# Applying the function
data.loc[data['Price Per Unit'].isna(),'Price Per Unit'] = (
data.loc[data['Price Per Unit'].isna(),'Item'].map(price_to_item))

In [49]:
data[data['Quantity'].isnull()].head(20)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
236     TXN_8562645     Salad       NaN             5.0          NaN   
278     TXN_3229409     Juice       NaN             3.0          NaN   
629     TXN_9289174      Cake       NaN             3.0         12.0   
641     TXN_2962976     Juice       NaN             3.0          NaN   
738     TXN_8696094  Sandwich       NaN             4.0          NaN   
912     TXN_1575608  Sandwich       NaN             4.0         20.0   
1008    TXN_7225428       Tea       NaN             1.5          3.0   
1436    TXN_7590801       Tea       NaN             1.5          6.0   
1482    TXN_3593060  Smoothie       NaN             4.0         16.0   
2330    TXN_3849488     Salad       NaN             5.0          5.0   
2796    TXN_9188692      Cake       NaN             3.0          NaN   
3203    TXN_4565754  Smoothie       NaN             4.0          NaN   
3224    TXN_6297232    Coffee       NaN             2.0          NaN   
3401    TXN_3251829       Tea       NaN             1.5          NaN   
3635    TXN_6177081    Cookie       NaN             1.0          1.0   
3779    TXN_7376255       NaN       NaN             NaN         25.0   
4257    TXN_6470865    Coffee       NaN             2.0          NaN   
5639    TXN_6206792       Tea       NaN             1.5          6.0   
5841    TXN_5884081    Cookie       NaN             1.0          NaN   
5845    TXN_8388462  Smoothie       NaN             4.0          8.0   

      Payment Method  Location Transaction Date  
236              NaN  In-store       2023-05-18  
278             Cash  Takeaway       2023-04-15  
629   Digital Wallet  In-store       2023-12-30  
641              NaN       NaN       2023-03-17  
738              NaN  Takeaway       2023-05-14  
912              NaN  Takeaway       2023-01-05  
1008     Credit Card  Takeaway       2023-03-07  
1436            Cash  Takeaway              NaT  
1482            Cash       NaN       2023-03-05  
2330             NaN  In-store       2023-03-01  
2796     Credit Card       NaN       2023-12-01  
3203  Digital Wallet  Takeaway       2023-10-06  
3224             NaN       NaN       2023-04-07  
3401  Digital Wallet  In-store       2023-07-25  
3635            Cash       NaN       2023-07-26  
3779             NaN  In-store       2023-05-27  
4257  Digital Wallet  Takeaway       2023-09-18  
5639     Credit Card       NaN       2023-10-13  
5841  Digital Wallet  In-store       2023-07-05  
5845             NaN  Takeaway       2023-08-19

In [51]:
data.loc[data['Total Spent'].isna(), 'Total Spent'] = data['Quantity'] * data['Price Per Unit']

In [69]:
data[data['Quantity'].isnull()].head(20)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
236     TXN_8562645     Salad       NaN             5.0          NaN   
278     TXN_3229409     Juice       NaN             3.0          NaN   
641     TXN_2962976     Juice       NaN             3.0          NaN   
738     TXN_8696094  Sandwich       NaN             4.0          NaN   
2796    TXN_9188692      Cake       NaN             3.0          NaN   
3203    TXN_4565754  Smoothie       NaN             4.0          NaN   
3224    TXN_6297232    Coffee       NaN             2.0          NaN   
3401    TXN_3251829       Tea       NaN             1.5          NaN   
3779    TXN_7376255       NaN       NaN             NaN         25.0   
4257    TXN_6470865    Coffee       NaN             2.0          NaN   
5841    TXN_5884081    Cookie       NaN             1.0          NaN   
7029    TXN_4628338    Coffee       NaN             2.0          NaN   
7297    TXN_9944500  Smoothie       NaN             4.0          NaN   
7597    TXN_1082717       NaN       NaN             NaN          9.0   
8021    TXN_2428781     Salad       NaN             5.0          NaN   
8443    TXN_2023651  Sandwich       NaN             4.0          NaN   
8465    TXN_9669616    Coffee       NaN             2.0          NaN   
8479    TXN_1547245  Sandwich       NaN             4.0          NaN   
8574    TXN_2546684     Juice       NaN             3.0          NaN   
8732    TXN_4550558    Cookie       NaN             1.0          NaN   

      Payment Method  Location Transaction Date  
236              NaN  In-store       2023-05-18  
278             Cash  Takeaway       2023-04-15  
641              NaN       NaN       2023-03-17  
738              NaN  Takeaway       2023-05-14  
2796     Credit Card       NaN       2023-12-01  
3203  Digital Wallet  Takeaway       2023-10-06  
3224             NaN       NaN       2023-04-07  
3401  Digital Wallet  In-store       2023-07-25  
3779             NaN  In-store       2023-05-27  
4257  Digital Wallet  Takeaway       2023-09-18  
5841  Digital Wallet  In-store       2023-07-05  
7029            Cash       NaN       2023-12-25  
7297            Cash  In-store       2023-01-03  
7597  Digital Wallet  In-store       2023-12-13  
8021             NaN  In-store       2023-05-09  
8443            Cash  In-store       2023-05-25  
8465             NaN       NaN       2023-06-03  
8479             NaN  Takeaway       2023-09-11  
8574  Digital Wallet  Takeaway       2023-04-08  
8732     Credit Card  In-store       2023-08-04

In [65]:
data.loc[
    data['Quantity'].isna() & data['Price Per Unit'].notna() & data['Total Spent'].notna(),
    'Quantity'
] = data['Total Spent'] / data['Price Per Unit']

 * Filled missing values in Quantity, Price Per Unit, and Total Spent using the values available.  
 * This ensures all valid values are captured and the dataset is consistent for calculations.  
Next Step  
Drop any remaining null values in the columns Quantity, Price Per Unit, and Total Spent to ensure a fully clean dataset.

In [74]:
# Drop rows where any of these 3 columns are null
data = data.dropna(subset=['Quantity', 'Price Per Unit', 'Total Spent'])

In [86]:
# Replace nulls in Payment Method and Location with "Unknown"
data['Payment Method'] = data['Payment Method'].fillna('Unknown')
data['Location'] = data['Location'].fillna('Unknown')
data['Item'] = data['Item'].fillna('Unknown')
# Replace nulls in Transaction data with 1900-01-01
data['Transaction Date'] = data['Transaction Date'].fillna(pd.Timestamp("1900-01-01"))

In [98]:
# Save the cleaned dataset to CSV
data.to_csv("cleaned_data.csv", index=False)

Transaction ID    Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373  Coffee       2.0             2.0          4.0   
1    TXN_4977031    Cake       4.0             3.0         12.0   
2    TXN_4271903  Cookie       4.0             1.0          4.0   
3    TXN_7034554   Salad       2.0             5.0         10.0   
4    TXN_3160411  Coffee       2.0             2.0          4.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         Unknown   Unknown       2023-04-27  
4  Digital Wallet  In-store       2023-06-11

In [88]:
data.isnull().sum()

Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64

In [96]:
data['Transaction Date'].value_counts()

Transaction Date
1900-01-01    460
2023-02-06     40
2023-06-16     40
2023-07-21     39
2023-09-21     39
             ... 
2023-04-27     15
2023-11-24     15
2023-03-11     14
2023-02-17     14
2023-07-22     14
Name: count, Length: 366, dtype: int64

In [92]:
data[['Item','Price Per Unit']].value_counts()

Item      Price Per Unit
Coffee    2.0               1286
Salad     5.0               1270
Cookie    1.0               1211
Tea       1.5               1206
Juice     3.0               1168
Cake      3.0               1138
Sandwich  4.0               1127
Smoothie  4.0               1094
Unknown   3.0                247
          4.0                227
Name: count, dtype: int64